### PR TITLE
Algo: Add timeseries forcasting model training script

### PR DESCRIPTION
**Pull request by dayko2019**
#### What type of PR is this?

/kind feature
/kind algorithm

#### What this PR does / why we need it:

Add time series prediction model training scripts, users can generate models with one click

#### Which issue(s) this PR fixes:

Fixes #

### CODE REVIEW

File: algorithm/kapacity/timeseries/forecasting/train.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+    # 4. train the model using data sets
+    train(args, config, df)
+
+    return
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(description='args of timeseries forcasting training')
+    parser.add_argument('--train-data-file', help='training data set file', required=False)
```
Comment (zqzten) [inline algorithm/kapacity/timeseries/forecasting/train.py:44]:
```suggestion
    parser.add_argument('--dataset-file', help='path of training dataset file, if set, will load dataset from this file instead of fetching from metrics server', required=False)
```

File: algorithm/kapacity/timeseries/forecasting/train.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    parser.add_argument('--train-config-file', help='training profile', required=False)
```
Comment (zqzten) [inline algorithm/kapacity/timeseries/forecasting/train.py:45]:
```suggestion
    parser.add_argument('--config-file', help='path of training config file', required=True)
```

File: algorithm/kapacity/timeseries/forecasting/train.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+import kapacity.portrait.horizontal.predictive.main as pred_main
+import kapacity.timeseries.forecasting.forecaster as forecaster
+
+
+def main():
+    # 1. parse arguments
+    args = parse_args()
+
+    # 2. fetch train config
```
Comment (zqzten) [inline algorithm/kapacity/timeseries/forecasting/train.py:30]:
```suggestion
    # 2. load training config
```

File: algorithm/kapacity/timeseries/forecasting/train.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    config = load_config(args.train_config_file)
+
+    # 3. fetch training data set
```
Comment (zqzten) [inline algorithm/kapacity/timeseries/forecasting/train.py:33]:
```suggestion
    # 3. fetch training dataset
```

File: algorithm/kapacity/timeseries/forecasting/train.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    df = fetch_train_data(args, config)
+
+    # 4. train the model using data sets
```
Comment (zqzten) [inline algorithm/kapacity/timeseries/forecasting/train.py:36]:
```suggestion
    # 4. train the model
```

File: algorithm/kapacity/timeseries/forecasting/train.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+    return
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(description='args of timeseries forcasting training')
+    parser.add_argument('--train-data-file', help='training data set file', required=False)
+    parser.add_argument('--train-config-file', help='training profile', required=False)
+    parser.add_argument('--metrics-server-addr', help='address of the gRPC metrics provider server', required=False)
+    parser.add_argument('--model-path', help='The directory where the model is stored', required=True)
```
Comment (zqzten) [inline algorithm/kapacity/timeseries/forecasting/train.py:47]:
```suggestion
    parser.add_argument('--model-save-path', help='dir path where the model and its related files would be saved in', required=True)
```

File: algorithm/kapacity/timeseries/forecasting/train.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+def train(args, config, df):
+    return forecaster.fit(
+        df=df,
+        freq=config['freq'],
+        target_col='value',
+        time_col='timestamp',
+        series_cols=['workload', 'metric'],
+        prediction_length=config['predictionLength'],
+        context_length=config['contextLength'],
+        learning_rate=config['contextLength'],
```
Comment (zqzten) [inline algorithm/kapacity/timeseries/forecasting/train.py:62]:
```suggestion
        learning_rate=config['hyperParams']['learningRate'],
```

File: algorithm/kapacity/timeseries/forecasting/train.py [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+    history_in_minutes = pred_main.time_period_to_minutes(config['historyLength'])
+    ago = (now - datetime.timedelta(minutes=history_in_minutes))
+    start = timestamp_pb2.Timestamp()
+    start.FromSeconds(int(ago.timestamp()))
+    end = timestamp_pb2.Timestamp()
+    end.FromSeconds(int(now.timestamp()))
+    return start, end
+
+
+def fetch_traffic_metrics(args, target):
```
Comment (zqzten) [inline algorithm/kapacity/timeseries/forecasting/train.py:96]:
```suggestion
def fetch_history_metrics(args, target):
```

File: algorithm/kapacity/timeseries/forecasting/train.py [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+    df_traffic = pd.DataFrame(columns=['timestamp', 'value', 'workload', 'metric'])
+    start, end = compute_history_range(target)
+
+    for j in range(len(target['metrics'])):
```
Comment (zqzten) [inline algorithm/kapacity/timeseries/forecasting/train.py:100]: just use `i` here

File: algorithm/kapacity/timeseries/forecasting/train.py [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+
+        if metric['type'] == 'Object':
+            df = pred_main.fetch_object_metric_history(args=args,
+                                                       namespace=workload_namespace,
+                                                       metric=metric,
+                                                       start=start,
+                                                       end=end)
+
+            df['metric'] = metric['object']['metric']['name']
+
```
Comment (zqzten) [inline algorithm/kapacity/timeseries/forecasting/train.py:113]: remove this empty line here

File: algorithm/kapacity/timeseries/forecasting/train.py [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+    ago = (now - datetime.timedelta(minutes=history_in_minutes))
+    start = timestamp_pb2.Timestamp()
+    start.FromSeconds(int(ago.timestamp()))
+    end = timestamp_pb2.Timestamp()
+    end.FromSeconds(int(now.timestamp()))
+    return start, end
+
+
+def fetch_traffic_metrics(args, target):
+    df_traffic = pd.DataFrame(columns=['timestamp', 'value', 'workload', 'metric'])
```
Comment (zqzten) [inline algorithm/kapacity/timeseries/forecasting/train.py:97]:
```suggestion
    df = pd.DataFrame(columns=['timestamp', 'value', 'workload', 'metric'])
```

File: algorithm/kapacity/timeseries/forecasting/train.py [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+    start, end = compute_history_range(target)
+
+    for j in range(len(target['metrics'])):
+        metric = target['metrics'][j]
+        workload_namespace = target['workloadNamespace']
+        workload_name = target['workloadName']
+
+        if metric['type'] == 'Object':
+            df = pred_main.fetch_object_metric_history(args=args,
```
Comment (zqzten) [inline algorithm/kapacity/timeseries/forecasting/train.py:106]:
```suggestion
            df_metric = pred_main.fetch_object_metric_history(args=args,
```

Comment (zqzten) [inline algorithm/kapacity/timeseries/forecasting/train.py:106]: suggest to extract common metrics query logic (including the `compute_history_range` stuff) to `kapacity.metric.query` package and avoid using other main package here

**Review by zqzten — CHANGES_REQUESTED** [review]

**Comment by zqzten** [comment]
replaced by #72
